content/posts/2018-10-21--complex-command-handler-in-javascript.md

```diff
@@ -0,0 +1,352 @@
+A3 - started, will take 1100 ms
+A3 - completed
+B1 - started, will take 1900 ms
+C1 - started, will take 300 ms
+D1 - started, will take 1300 ms
+C1 - completed
+C2 - started, will take 2100 ms
+D1 - completed
+D2 - started, will take 2100 ms
+B1 - completed
+B2 - started, will take 1900 ms
+C2 - completed
+C3a - started, will take 1100 ms
+C3b - started, will take 1200 ms
+C3c - started, will take 1800 ms
+C3d - started, will take 1300 ms
+C3e - started, will take 1300 ms
+D2 - completed
+C3a - completed
+C3b - completed
+C3d - completed
+C3e - completed
+B2 - completed
+B3 - started, will take 1800 ms
+C3c - completed
+C4 - started, will take 700 ms
+C4 - completed
+C5 - started, will take 400 ms
+C5 - completed
+C6 - started, will take 500 ms
+B3 - completed
+B4 - started, will take 2000 ms
+C6 - completed
+B4 - completed
+B5 - started, will take 1500 ms
+B5 - completed
+B6 - started, will take 1300 ms
+B6 - completed
+B7 - started, will take 900 ms
+B7 - completed
+E1 - started, will take 1000 ms
+E1 - completed
+E2 - started, will take 1300 ms
+E2 - completed
+E3 - started, will take 300 ms
+E3 - completed
+All commands completed.
+```
+
+So our code already does what it was expected to do, but the implementation of `handleInSequence` method is not as readable as we would like. Let's check if using `async`/`await` syntactic sugar could improve this.
+
+```js
+async handleInSequence(commands) {
+    for (const command of commands) {
+        await this.handle(command);
+    }
+}
+```
+
+Wow, that's something! As you can see we only need to loop through provided commands and `await` for each `handle` call. This will pause the execution of this method until this nested promise is resolved. (Of course it won't stop the execution of other callbacks that are in motion as this would beat the purpose of using asynchronous tool like Node.js). Since we don't care about results returned we can end our method without explicitly returning any value --- defining this method as `async` will automatically wrap it's result value with `Promise` object that caller method expects to receive!
+
+We can clearly see that this implementation is better and easier to analyze, so let's use it instead of the previous one.
+
+Since it went so well for `handleInSequence`, let's now get back to implementing `handleInParallel` but this time using `async`/`await` approach. In this case it doesn't result in an easy to read code, though.
+
+First we can wonder why we can't simply call `handle` method for each command inside a `for ... of` loop and `await` for them to be resolved? The problem is that this will not yield the expected result --- note that we've already used this approach when we implemented `handleInSequence` method! The reason is that `await` would pause the execution of our method for every single `handle` call. This would prevent handling of subsequent commands for even getting started and instead of being executed in parallel they would be run in sequence.
```

Comment:
```suggestion
First, we can wonder why we can't simply call the `handle` method for each command inside a `for ... of` loop and `await` for them to be resolved? The problem is that this will not yield the expected result --- note that we've already used this approach when we implemented `handleInSequence` method! The reason is that `await` would pause the execution of our method for every single `handle` call. This would prevent handling of subsequent commands for even getting started and instead of being executed in parallel they would be run in sequence.
```